File: medium/Unique Length-3 Palindromic Subsequences/main.cpp
// 1930. Unique Length-3 Palindromic Subsequences
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        int n = s.size();
        vector<int> first(26, -1), last(26, -1);

        for (int i = 0; i < n; i++) {
            int c = s[i] - 'a';
            if (first[c] == -1) first[c] = i;
            last[c] = i;
        }

        int ans = 0;

        for (int a = 0; a < 26; a++) {
            int l = first[a], r = last[a];
            if (l != -1 && r - l > 1) {
                bool seen[26] = {0};
                for (int i = l + 1; i < r; i++) {
                    int b = s[i] - 'a';
                    if (!seen[b]) {
                        seen[b] = true;
                        ans++;
                    }
                }
            }
        }
        return ans;
    }
};

File: medium/Unique Length-3 Palindromic Subsequences/main.rs
// 1930. Unique Length-3 Palindromic Subsequences
impl Solution {
    pub fn count_palindromic_subsequence(s: String) -> i32 {
        let bytes = s.as_bytes();
        let n = bytes.len();

        let mut first = vec![-1; 26];
        let mut last = vec![-1; 26];

        for (i, &c) in bytes.iter().enumerate() {
            let idx = (c - b'a') as usize;
            if first[idx] == -1 {
                first[idx] = i as i32;
            }
            last[idx] = i as i32;
        }

        let mut ans = 0;

        for a in 0..26 {
            let l = first[a];
            let r = last[a];
            if l != -1 && r - l > 1 {
                let mut seen = [false; 26];
                for i in (l + 1) as usize..r as usize {
                    let b = (bytes[i] - b'a') as usize;
                    if !seen[b] {
                        seen[b] = true;
                        ans += 1;
                    }
                }
            }
        }

        ans
    }
}